lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,985 @@
+          auto dstOldType = partial.dest().getType();
+          auto dstNewType = dstOldType;
+          auto srcOldType = partial.src().getType();
+          auto srcNewType = srcOldType;
+          // Check which side of the partial connect was updated.
+          if (newValue == partial.dest())
+            dstOldType = oldType;
+          else
+            srcOldType = oldType;
+          ImplicitLocOpBuilder builder(partial.getLoc(), partial);
+          fixupPartialConnect(
+              builder, [&](auto) { return partial.dest(); }, dstNewType,
+              dstOldType, [&](auto) { return partial.src(); }, srcNewType,
+              srcOldType);
+          partial->erase();
+        } else if (auto connect = dyn_cast<ConnectOp>(op)) {
+          // Rewrite the connect to connect all values.
+          auto dst = connect.dest();
+          auto src = connect.src();
+          ImplicitLocOpBuilder builder(connect.getLoc(), connect);
+          // Check which side of the connect was updated.
+          if (newValue == dst)
+            fixupConnect(builder, dst, oldType, src, src.getType());
+          else
+            fixupConnect(builder, dst, dst.getType(), src, oldType);
+          connect->erase();
+        }
+      }
+    }
+  }
+
+  /// This is the root method to fixup module references when a module changes.
+  /// It matches all the results of "to" module with the results of the "from"
+  /// module.
+  void fixupReferences(FModuleLike toModule, Operation *fromModule) {
+    // Replace all instances of the other module.
+    auto *node = instanceGraph[fromModule];
+    for (auto *use : node->uses()) {
+      auto oldInst = use->getInstance();
+      // Create an instance to replace the old module.
+      auto newInst = OpBuilder(oldInst).create<InstanceOp>(
+          oldInst.getLoc(), toModule, oldInst.nameAttr());
+      newInst.annotationsAttr(oldInst.annotationsAttr());
+      auto oldInnerSym = oldInst.inner_symAttr();
+      if (oldInnerSym)
+        newInst.inner_symAttr(oldInnerSym);
+      // We have to replaceAll before fixing up references, we walk the new
+      // usages when fixing up any references.
+      oldInst.replaceAllUsesWith(newInst.getResults());
+      // Update bulk connections and subfield operations.
+      for (auto results :
+           llvm::zip(newInst.getResults(), oldInst.getResultTypes()))
+        fixupReferences(std::get<0>(results), std::get<1>(results));
+      oldInst->erase();
+    }
+  }
+
+  /// Private NLA creation method.  This function requires an attribute array
+  /// with a placeholder in the first element, where the root refence of the NLA
+  /// will be inserted. This clones the NLA for each instantiation of the
+  /// "fromModule".
+  SmallVector<FlatSymbolRefAttr>
+  createNLAsImpl(OpBuilder &builder, StringAttr toModuleName, AnnoTarget to,
+                 Operation *fromModule, SmallVectorImpl<Attribute> &namepath) {
```

Comment:
Nit: should this create the NLAs at the top of the circuit as opposed to wherever the `builder` happens to be?

Reply:
I was trying to place them above the module that stays when two modules dedup, but it ends up looking random because i wrote the test modules in leaf-first order.  If i write the parent module first it looks a little cleaner.  I agree that this does seem a little random right now, do we have a preference for putting all the NLAs at the top of the circuit?   If so I guess I can put each test in its own circuit, otherwise I can swap the order of modules so it looks better.

Reply:
It seems cleaner to aggregate them at the top of the circuit or at least before the first module that uses them. This is not a hard requirement, though.

(This also came from me writing tests... I had expected to the see the NLA op show up before any symbol uses, but it wasn't the case.)

Not a big deal.

Reply:
Fixed this up to put the NLAs at the top of the circuit. I ran into a problem forcing me to insert each NLA right at the beginning of the circuit, effectively leaving them in the reverse order they were created in.